dynamicio/io/s3.py

```diff
@@ -0,0 +1,82 @@
+from functools import partial
+from pathlib import Path
+from typing import Any, Dict, List, Literal, Optional, Type
+
+import boto3
+import pandas as pd
+
+from dynamicio.io.s3_contexts import s3_named_file_reader, s3_writer, s3_reader
+from dynamicio.io.resource import BaseResource
+from dynamicio.io.serde import BaseSerde, CsvSerde, HdfSerde, JsonSerde, ParquetSerde, PickleSerde
+
+
+class S3Resource(BaseResource):
+    bucket: str
+    path: Path
+    read_kwargs: Dict[str, Any] = {}
+    write_kwargs: Dict[str, Any] = {}
+    injectables: List[str] = ["path"]
+    file_type: Optional[Literal["parquet", "hdf", "csv", "json", "pickle"]] = None
+    force_read_to_memory: bool = False
+
+    @property
+    def _s3_path(self) -> str:
+        """For logging purposes only."""
+        return f"s3://{self.bucket}/{self.path}"
+
+    def _read(self) -> pd.DataFrame:
+        if self.force_read_to_memory:
+            with s3_reader(boto3.client("s3"), s3_bucket=self.bucket, s3_key=str(self.path)) as fobj:  # type: ignore
+                df = self.get_serde()._read(fobj, **self.read_kwargs)  # type: ignore
+                if df is not None:
+                    return df
+
+        with s3_named_file_reader(boto3.client("s3"), s3_bucket=self.bucket, s3_key=str(self.path)) as target_file:
+            return self.get_serde()._read(target_file.name, **self.read_kwargs)  # type: ignore
+
+    def _write(self, df: pd.DataFrame) -> None:
+        with s3_writer(boto3.client("s3"), s3_bucket=self.bucket, s3_key=str(self.path)) as fobj:
+            return self.get_serde()._write(fobj, df)
+
+    @property
+    def serde_class(self):
```

Comment:
nit: given that both the `LocalFileResource` and the `S3Resource` use this implementation I wonder whether it may make sense to have this moved to the `BaseResource` class?

Reply:
It is in the base resource - the default being the pickleserde.

Reply:
If we want any class to have file_type which would by default refer to the serde we could do that, why not.